clang/lib/3C/ProgramInfo.cpp

```diff
@@ -649,10 +668,16 @@ void ProgramInfo::unifyIfTypedef(const Type* Ty, ASTContext& Context, Declarator
 
 bool ProgramInfo::hasPersistentConstraints(Expr *E, ASTContext *C) const {
   auto PSL = PersistentSourceLoc::mkPSL(E, *C);
+  bool HasImpCastConstraint =
+    isa<ImplicitCastExpr>(E) &&
+    ImplicitCastConstraintVars.find(PSL) != ImplicitCastConstraintVars.end() &&
+    !ImplicitCastConstraintVars.at(PSL).empty();
+  bool HasExprConstraint =
+    !isa<ImplicitCastExpr>(E) &&
+    ExprConstraintVars.find(PSL) != ExprConstraintVars.end() &&
+    !ExprConstraintVars.at(PSL).empty();
   // Has constraints only if the PSL is valid.
-  return PSL.valid() &&
-         ExprConstraintVars.find(PSL) != ExprConstraintVars.end() &&
-         !ExprConstraintVars.at(PSL).empty();
+  return PSL.valid() && (HasExprConstraint || HasImpCastConstraint);
```

Comment:
Similar question to earlier: Here we are checking for casts on internal *or* external constraints. Why not one or the other? Depends on how this function is used?